src/main/java/searching/Tree23.java

```diff
@@ -0,0 +1,261 @@
+ */
+public class Tree23 {
+
+    Node23 root;
+
+    public Tree23(){
+        root = null;
+    }
+
+    static Node23 carry;
+    static Node23 subtree, subtree2;
+
+    public Node23 put(Node23 n, Integer key, Double value){
+        if (n == null) return new Node23(key, value, null, null);
+        Node23 ret = n.put(key, value);
+        if (carry != null) {
+            carry.left = subtree;
+            carry.middle = subtree2;
+            return carry;
+        }
+        return ret;
+    }
+
+    public void put(Integer key, Double value){
+        // BEGIN STRIP
+        carry = null;
+        root = put(root, key, value);
+        // END STRIP
+        // TODO
+        // STUDENT return;
+
+    }
+
+    public Double search(Integer k){
+        if (root == null) return null;
+        Double temp = root.search(k);
+        return temp;
+    }
+
+    public boolean isBalanced(){
+        if (root == null) return true;
+        return root.isBalanced() >= 0;
+    }
+
+    public boolean isBst(){
+        if (root == null) return true;
+        return root.isBst(Integer.MIN_VALUE, Integer.MAX_VALUE);
+    }
+    
+}
+
+class Node23{
+
+    Integer k1, k2;  // k1 < k2
+    Double v1, v2;
+    Node23 left, middle, right;
+    // when we have a 2Node => only left and middle are used and k2 is set to INF
+    // thus any key greater than k1 will be less than k2 (used for search for example)
+
+    Node23(Integer k, Double v, Node23 left, Node23 right){
+        this.k1 = k;
+        this.v1 = v;
+        this.left = left;
+        this.middle = right; // is comprised between [k1, inf[
+        this.k2 = Integer.MAX_VALUE;
+        this.right = null;
+    }
+
+    Node23(Integer k1, Integer k2, Double v1, Double v2, Node23 left, Node23 middle, Node23 right){
+        this.k1 = Math.min(k1, k2);
+        this.k2 = Math.max(k1, k2);
+        this.v1 = k1<k2 ? v1 : v2;
+        this.v2 = k1<k2 ? v2 : v1;
+        this.left = left;
+        this.right = right;
+        this.middle = middle;
+    }
+
+    public boolean is2Node(){ return k2==Integer.MAX_VALUE; }
+    public boolean is3Node(){ return !is2Node(); }
+
+    /**
+     *
+     * @param k : a key to compare
+     * @return 0 if k1 < k < k2
+     *        -1 if k < k1
+     *         1 if k > k2
+     */
+    public int compare(Integer k){
+        if (k.compareTo(k1) > 0 && k.compareTo(k2) < 0) return 0;
+        if (k.compareTo(k1) < 0 ) return -1;
+        return 1;
+    }
+
+    public boolean isLeaf() {
+        return left == null && middle == null && right == null;
+    }
+
+    public Double search(Node23 n, Integer k) {
+        if (n==null) return null;
+        if (n.k1.equals(k)) return n.v1;
+        if (n.k2.equals(k)) return n.v2;
+
+        if (k.compareTo(n.k1)>0 && k.compareTo(n.k2) <0) return search(n.middle, k);
+        else if (k.compareTo(n.k1) < 0) return search(n.left, k);
+        else return search(n.right, k);
+    }
+
+    public Double search(Integer k){
```

Comment:
Comments.
Also `temp` variable is useless in this case, just return the call to the other function